### ISSUE TITLE
Delivery price cannot be defined by shop

### ISSUE THREAD

**Issue by jf-viguier** [PrestaShop/PrestaShop#10383]
**Describe the bug**
In multishop context, it's impossible to configure a delivery price for a specific shop. The context is forced to "All shops".

**To Reproduce**
1. Select a shop in multoshop context
2. Go to the add carrier wizard 
3. The shop dropdown is locked to "All shops"

**Additionnal information**
PrestaShop version: all 1.6 ans 1.7
PHP version: 5.6

**Comment by MathiasReker** [comment]
PR: https://github.com/PrestaShop/PrestaShop/pull/10384

**Comment by khouloudbelguith** [comment]
Hi @jf-viguier,

Yes, you are right.
Thanks for your report & your fix.

Best regards, KHouloud

**Comment by BilalDr** [comment]
Hi, @jf-viguier after checking your issue i found that on some pages of the BO when multistore is activated, you can't select a context and it's forced on All shops because on those pages there is a 
 "Multistore" section  where you can select store to configure for example a delivery price for a specific shop .

![image](https://user-images.githubusercontent.com/40759259/46745797-12684800-ccae-11e8-8dd5-05e82e74ccce.png)

**Comment by khouloudbelguith** [comment]
Hi,

Thanks, @BilalDr!
We need to specify this issue.
For example in the category page, there is a "Multistore" section where you can select store, but we can select a context and it's not forced on All shops
![image](https://user-images.githubusercontent.com/16067358/46746455-66276100-ccaf-11e8-959c-3e4334b9fd83.png)
@BilalDr what do you think?
Thanks!

**Comment by jf-viguier** [comment]
The "Multistore" section is usefull to setup if the carrier is available for a specific shop but my concern is to define per shop the carrier prices or ranges.

**Comment by BilalDr** [comment]
@khouloudbelguith Indeed, sometimes, there is context available and Shop Association section. 
On this case, Shop association is used to define where you want to show your element (here the category "Art") and you have to be on the context All shops to have an overview.

@jf-viguier , If you want to have specific price for your carrier on your shop, you just have to select your shop on step 2 "MultiStore" and on the step 3 "Shipping locations and costs" you will be able to define prices for your carrier on the shop selected previously 

Is it what you wanted ? @jf-viguier

**Comment by jf-viguier** [comment]
No. I want to define for THE SAME carrier a specific delivery price par shop without creating a new carrier for each shop.

**Comment by BilalDr** [comment]
@jf-viguier then if you want that it isn't currently possible.

**Comment by jf-viguier** [comment]
It's possible with my pr #10384 but this has to be tested.

**Comment by sarahdib** [comment]
@jf-viguier @eternoendless @PierreRambaud 

With @marionf we found some error 

https://drive.google.com/file/d/1Os0sltAg8CGIatdmsvfO9WNuDkHtCX4n/view

 1- When you have two shop (shop A and shop B)  
When we are In shop A context 
-> add new carrier 
-> set up the information in first tab and click on next
-> Multistore Tab : all shops (all chackbox) are selected instead of shop A only(see the behavior in supplier)

2- Edit a carrier in shop A change the context to shop B and add shop B to the shop autorize to use the carrier.
In the last step when we click on finish there is an error

3- Create a carrier in context all shops
-> Set up the carrier and validate
-> Switch to shop A context and change the price. validate the change 
-> Go on carrier listing
-> Switch to shop B context 
-> Edit the carrier and see that the price who was set up for all shop has been removed
-> Set up a new price for shop B. validate the change 
-> Go on carrier listing
-> Switch to shop A 
-> Edit the carrier and edit the carrier there is no price anymore 
https://drive.google.com/file/d/10UmxD3XvD_Mw7IPHJVOnIfR0RESTfZD3/view

**Comment by boo-code** [comment]
Still true on `develop`, but the mechanism has moved, so anyone picking this up from the description
above would patch the wrong file.

`AdminCarrierWizardController` still carries `$this->multishop_context = Shop::CONTEXT_ALL;`, and it
is no longer the page that saves a carrier: the add/edit page has been migrated, to
`PrestaShopBundle\Controller\Admin\Improve\Shipping\CarrierController` behind
`admin_carriers_create` and `admin_carriers_edit`. The limitation came across the migration and is
explicit in the new code. `Adapter\Carrier\Repository\CarrierRangeRepository` writes delivery rows as

    'price' => $range->getPrice(),
    // Only all shops is handled for now
    'id_shop' => null,
    'id_shop_group' => null,

with the same two hardcoded nulls again in `reset()`, and its `assertShopConstraint()` refuses
anything else outright:

    if (!$shopConstraint->forAllShops()) {
        throw new CarrierConstraintException('Shop constraint isn\'t supported yet.', ...);
    }

Worth recording because the storage and the read path are both ready for it. `delivery` has
`id_shop` and `id_shop_group` columns, and `Carrier::sqlDeliveryRangeShop()` already resolves them
with shop-over-group-over-global precedence:

    ORDER BY d2.id_shop DESC, d2.id_shop_group DESC
    LIMIT 1

selecting `d2.id_shop = <context shop>` first, then the group row, then the row with both columns
null. On a stock shop all 16 `delivery` rows have `id_shop` null, which is what that last branch
resolves.

So the read side implements per-shop delivery prices and only the write side is pinned to all shops -
deliberately, per the comments in the migrated repository, rather than by the legacy context call
this report names.
